src/Microsoft.Diagnostics.Monitoring.StartupHook/Exceptions/Pipeline/ExceptionPipeline.cs

```diff
@@ -31,15 +32,19 @@ public ExceptionPipeline(ExceptionSourceBase exceptionSource, Action<ExceptionPi
         public void Start()
         {
             DisposableHelper.ThrowIfDisposed<ExceptionPipeline>(ref _disposedState);
+            _processingState = 0;
 
             _exceptionSource.ExceptionAvailable += ExceptionSource_ExceptionAvailable;
         }
 
         public void Stop()
         {
+            // We must do this first to prevent any new exception callbacks.
             _exceptionSource.ExceptionAvailable -= ExceptionSource_ExceptionAvailable;
 
-            while (0 != Interlocked.CompareExchange(ref _processingState, 0, 0) && !DisposableHelper.IsDisposed(ref _disposedState))
+            // Wait until all outstanding exception callbacks are finished. Once _processingState reaches 0, we will set the WriteFlag.
+            // This will prevent any further exception handlers from processing.
+            while (0 != Interlocked.CompareExchange(ref _processingState, WriteFlag, 0) && !DisposableHelper.IsDisposed(ref _disposedState))
             {
                 // Wait for all exceptions to be processed
                 Thread.Sleep(100);
```

Comment:
Should SpinWait be used here too?

Reply:
I purposely did not. My thinking is that there are N threads that could be throwing exceptions and I want those to have as little contention as possible and I don't want to put the thread to sleep if possible.  The Stop thread on the other hand should not be eating cpu if there are many outstanding exceptions being processed. It should yield its timeslice liberally to the rest of the app. Starvation is not a concern because I first unhook the exception handler.